certbot-ci/certbot_integration_tests/utils/misc.py

```diff
@@ -30,16 +40,142 @@ def check_until_timeout(url):
 
 
+def find_certbot_root_directory():
+    script_path = os.path.realpath(__file__)
+    current_dir = os.path.dirname(script_path)
+
+    while '.git' not in os.listdir(current_dir) and current_dir != os.path.dirname(current_dir):
+        current_dir = os.path.dirname(current_dir)
+
+    dirs = os.listdir(current_dir)
+    if '.git' not in dirs:
+        raise ValueError('Error, could not find certbot sources root directory')
+
+    return current_dir
+
+
+def generate_csr(domains, key_path, csr_path, key_type='RSA'):
+    certbot_root_directory = find_certbot_root_directory()
+    script_path = os.path.join(certbot_root_directory, 'examples', 'generate-csr.py')
+
+    command = [
+        sys.executable, script_path, '--key-path', key_path, '--csr-path', csr_path,
+        '--key-type', key_type
+    ]
+    command.extend(domains)
+    subprocess.check_call(command)
+
+
+def load_sample_data_path(workspace):
+    certbot_root_directory = find_certbot_root_directory()
+    original = os.path.join(certbot_root_directory, 'tests', 'integration', 'sample-config')
+    copied = os.path.join(workspace, 'sample-config')
+    shutil.copytree(original, copied, symlinks=True)
+    return copied
+
+
+def read_certificate(cert_path):
+    with open(cert_path, 'r') as file:
+        data = file.read()
+
+    cert = crypto.load_certificate(crypto.FILETYPE_PEM, data.encode('utf-8'))
+    return crypto.dump_certificate(crypto.FILETYPE_TEXT, cert).decode('utf-8')
+
+
+class GraceFullTCPServer(socketserver.TCPServer):
+    allow_reuse_address = True
+
+
 @contextlib.contextmanager
-def execute_in_given_cwd(cwd):
-    """
-    Context manager that will execute any command in the given cwd after entering context,
-    and restore current cwd when context is destroyed.
-    :param str cwd: the path to use as the temporary current workspace for python execution
-    """
+def create_tcp_server(port):
     current_cwd = os.getcwd()
+    webroot = tempfile.mkdtemp()
+
+    def run():
+        GraceFullTCPServer(('', port), SimpleHTTPServer.SimpleHTTPRequestHandler).serve_forever()
+
+    process = multiprocessing.Process(target=run)
+
     try:
-        os.chdir(cwd)
-        yield
+        try:
+            os.chdir(webroot)
+            process.start()
+        finally:
+            os.chdir(current_cwd)
+
+        check_until_timeout('http://localhost:{0}/'.format(port))
+
+        yield webroot
     finally:
-        os.chdir(current_cwd)
+        try:
+            if process.is_alive():
+                process.terminate()
+                process.join()  # Block until process is effectively terminated
+        finally:
+            shutil.rmtree(webroot)
+
+
+def list_renewal_hooks_dirs(config_dir):
+    renewal_hooks_root = os.path.join(config_dir, 'renewal-hooks')
+    return [os.path.join(renewal_hooks_root, item) for item in ['pre', 'deploy', 'post']]
+
+
+def generate_test_file_hooks(config_dir, hook_probe):
```

Comment:
Designed to be called at some point of a test, before calling `certbot`, to check that hooks scripts are correctly invoked. Theses files are typically deleted just after, and the test finishes and so the entire workspace is deleted.

Reply:
I can't say I love having to call out to a magic setup function before invoking certbot with hooks. In `unittest`, we'd put all of the renew functions in a class, and put this in the `setup` function of that class. Is there a similar way we could do that here?

Reply:
Yes, this is the purpose of a fixture.

Reply:
However, the setup, like the fixture, are designed to have a common logic to be executed for each test in a logic group (either as a method in a TestCase class, or as a group of functions sharing the same fixture in their parameters).

For a logic that is triggered in only one specific test, I do not see the gain to rely on a implicit setup from the Test Case or a fixture, versus a helper function, if the purpose is to abstract come logic out of the test code itself.

Reply:
To be, abstracting out to a fixture class and a helper functions class imply the same level of abstraction. The difference is that one has to do with setup, whereas one takes out significant code chunks. But anyway, I haven't looked fully yet, but expect that the `test_renew` might be able to be broken out into smaller functions, since it looks like there's some setup that happens multiple times within the same function.

Reply:
As it is related to the test_renew, that could take a significant leap of quality by refactoring, but that I would like to process in the next PR when we will be on the test code itself, I propose you to continue the discussion about this item in the next item.